File: Backend/demo/src/main/java/com/example/Services/InvoicePdfService.java
package com.example.Services;

import com.example.Models.BookingDetail;
import com.example.Models.BookingHeader;
import com.example.Models.AddOnMaster;
import com.lowagie.text.*;
import com.lowagie.text.pdf.PdfPCell;
import com.lowagie.text.pdf.PdfPTable;
import com.lowagie.text.pdf.PdfWriter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Optional;

@Service
public class InvoicePdfService {

    @Autowired
    private BookingHeaderService bookingHeaderService;

    @Autowired
    private BookingDetailService bookingDetailService;

    @Autowired
    private AddOnMasterService addOnMasterService;

    public byte[] generateInvoice(Long bookingId) throws Exception {
        Optional<BookingHeader> bookingOptional = bookingHeaderService.getBookingById(bookingId);
        if (!bookingOptional.isPresent()) {
            throw new RuntimeException("Booking not found with ID: " + bookingId);
        }
        BookingHeader booking = bookingOptional.get();

        List<BookingDetail> bookingDetails = bookingDetailService.getBookingDetailByBookingId(bookingId.intValue());

        Document document = new Document();
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        PdfWriter.getInstance(document, out);

        document.open();

        Font titleFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 16);
        Paragraph title = new Paragraph("Invoice", titleFont);
        title.setAlignment(Element.ALIGN_CENTER);
        document.add(title);

        document.add(new Paragraph("Booking ID: " + booking.getBookingId()));
        document.add(new Paragraph("Customer Name: " + booking.getFirstname() + " " + booking.getLastname()));
        document.add(new Paragraph("Email: " + booking.getEmailId()));
        document.add(new Paragraph("Booking Date: " + booking.getBookingdate()));

        PdfPTable table = new PdfPTable(4);
        table.setWidthPercentage(100);
        table.setSpacingBefore(10);
        table.addCell("Add-On ID");
        table.addCell("Add-On Name");
        table.addCell("Rate");
        table.addCell("Total");

        long rentalDays = ChronoUnit.DAYS.between(booking.getStartdate(), booking.getEnddate() ) +1;
        
        double dailyRate = booking.getDailyrate();
        double weeklyRate = booking.getWeeklyrate();
        double monthlyRate = booking.getMonthlyrate();
        
        long months = rentalDays / 30;
        long remainingDaysAfterMonths = rentalDays % 30;
        long weeks = remainingDaysAfterMonths / 7;
        long remainingDays = remainingDaysAfterMonths % 7;

        double totalBookingCost = (months * monthlyRate) + (weeks * weeklyRate) + (remainingDays * dailyRate);
        
        double totalAddonCost = 0;
        for (BookingDetail detail : bookingDetails) {
            AddOnMaster addOn = addOnMasterService.getAddOnById(detail.getAddonId());
            table.addCell(String.valueOf(addOn.getAddonId()));
            table.addCell(addOn.getAddonName());
            table.addCell("$" + addOn.getAddonDailyRate());
            double total = addOn.getAddonDailyRate() * rentalDays;

            table.addCell("$" + total);
            totalAddonCost += total;
        }
        document.add(table);

        double grandTotal = totalBookingCost + totalAddonCost;

        document.add(new Paragraph("Total Booking Cost: $" + totalBookingCost));
        document.add(new Paragraph("Total Add-On Cost: $" + totalAddonCost));
        document.add(new Paragraph("Grand Total: $" + grandTotal));

        document.close();
        return out.toByteArray();
    }
}

File: Dot net Backend/latest/Fleet_Management/Service/InvoicePdfService.cs
using System;
using System.IO;
using System.Threading.Tasks;
using iText.Kernel.Pdf;
using iText.Layout;
using iText.Layout.Element;
using iText.Layout.Properties;
using iText.IO.Font.Constants;
using iText.Kernel.Font;
using System.Text;
using Fleet_Management.Service;


namespace Fleet_Management.Service
{
    public class InvoicePdfService : IInvoicePdfService
    {
        private readonly IBookingHeaderService _bookingHeaderService;
        private readonly IBookingDetailService _bookingDetailService;
        private readonly AddOnMasterService _addOnMasterService;

        public InvoicePdfService(IBookingHeaderService bookingHeaderService,
                                 IBookingDetailService bookingDetailService,
                                 IAddOnMasterService addOnMasterService)
        {
            _bookingHeaderService = bookingHeaderService;
            _bookingDetailService = bookingDetailService;
            _addOnMasterService = (AddOnMasterService?)addOnMasterService;
        }

        public async Task<byte[]> GenerateInvoiceAsync(long bookingId)
        {
            var booking = await _bookingHeaderService.GetBookingByIdAsync(bookingId);
            if (booking == null)
                throw new Exception("Booking not found with ID: " + bookingId);

            var bookingDetails = await _bookingDetailService.GetBookingDetailsByBookingId((int)bookingId);

            using (var stream = new MemoryStream())
            {
                var writerProperties = new WriterProperties();

                // Encryption Setup
                var writer = new PdfWriter(stream, new WriterProperties()
            .SetStandardEncryption(
                Encoding.UTF8.GetBytes("userPassword"), 
                Encoding.UTF8.GetBytes("ownerPassword"), 
                EncryptionConstants.ALLOW_PRINTING, 
                EncryptionConstants.ENCRYPTION_AES_128 
            ));
                var pdf = new PdfDocument(writer);
                var document = new Document(pdf);

                var boldFont = PdfFontFactory.CreateFont(StandardFonts.HELVETICA_BOLD);
                var title = new Paragraph("Invoice")
                                .SetTextAlignment(TextAlignment.CENTER)
                                .SetFontSize(16)
                                .SetFont(boldFont);
                document.Add(title);

                document.Add(new Paragraph("Booking ID: " + booking.BookingId));
                document.Add(new Paragraph("Customer Name: " + booking.Firstname + " " + booking.Lastname));
                document.Add(new Paragraph("Email: " + booking.EmailId));
                document.Add(new Paragraph("Booking Date: " + booking.Bookingdate));

                var table = new Table(4)
                                .UseAllAvailableWidth()
                                .SetMarginTop(10);
                table.AddHeaderCell("Add-On ID");
                table.AddHeaderCell("Add-On Name");
                table.AddHeaderCell("Rate");
                table.AddHeaderCell("Total");

                var rentalDays = (booking.Enddate.ToDateTime(TimeOnly.MinValue) -
                                  booking.Startdate.ToDateTime(TimeOnly.MinValue)).Days + 1;

                double dailyRate = booking.Dailyrate ?? 0;
                double weeklyRate = booking.Weeklyrate ?? 0;
                double monthlyRate = booking.Monthlyrate ?? 0;

                long months = rentalDays / 30;
                long remainingDaysAfterMonths = rentalDays % 30;
                long weeks = remainingDaysAfterMonths / 7;
                long remainingDays = remainingDaysAfterMonths % 7;

                double totalBookingCost = (months * monthlyRate) + (weeks * weeklyRate) + (remainingDays * dailyRate);

                double totalAddonCost = 0;
                foreach (var detail in bookingDetails)
                {
                    var addOn = await _addOnMasterService.GetAddOnByIdAsync(detail.AddonId ?? 0);
                    if (addOn != null)
                    {
                        table.AddCell(addOn.AddonId.ToString());
                        table.AddCell(addOn.AddonName);
                        table.AddCell("$" + addOn.AddonDailyRate);
                        double total = (addOn.AddonDailyRate ?? 0) * rentalDays;
                        table.AddCell("$" + total);
                        totalAddonCost += total;
                    }
                }
                document.Add(table);

                double grandTotal = totalBookingCost + totalAddonCost;

                document.Add(new Paragraph("Total Booking Cost: $" + totalBookingCost));
                document.Add(new Paragraph("Total Add-On Cost: $" + totalAddonCost));
                document.Add(new Paragraph("Grand Total: $" + grandTotal));

                document.Close();
                return stream.ToArray();
            }
        }
    }
}